File: code-examples/example-1/src/test/java/be/swsb/typesafety/example1/rest/BookingResourceTest.java
package be.swsb.typesafety.example1.rest;

import be.swsb.typesafety.example1.rest.booking.BookingResource;
import be.swsb.typesafety.example1.service.booking.BookingService;
import be.swsb.typesafety.test.UnitTest;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import static org.mockito.Mockito.verify;

class BookingResourceTest extends UnitTest {

    @InjectMocks
    private BookingResource resource;

    @Mock
    private BookingService service;

    @Test
    void saveBooking_ShouldCreateANewBookingAndReturnItsUUID() {
        final String bookingCode = "12345678909876543";

        resource.saveBooking(bookingCode);

        verify(service).createNewBooking(bookingCode);
    }
}

File: code-examples/examples/example-5/src/test/java/be/swsb/typesafety/example5/rest/booking/BookingResourceTest.kt
package be.swsb.typesafety.example5.rest.booking

import be.swsb.typesafety.example5.domain.booking.Booking
import be.swsb.typesafety.example5.domain.booking.BookingCode
import be.swsb.typesafety.example5.service.booking.BookingService
import be.swsb.typesafety.test.UnitTest
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*

internal class BookingResourceTest : UnitTest() {

    private lateinit var resource: BookingResource

    private val service: BookingService = mock()

    @BeforeEach
    internal fun setUp() {
        resource = BookingResource(service)
    }

    @Test
    fun saveBooking_ShouldCreateANewBookingAndReturnItsUUID() {
        val bookingCode = "12345678"

        val newUUID = UUID.randomUUID()
        whenever(service.createNewBooking(BookingCode.bookingCode(bookingCode))).thenReturn(newUUID)

        val actual = resource.saveBooking(bookingCode)

        assertThat(actual).isEqualTo(newUUID)
    }

    @Test
    fun chainingOptionalValues() {
        val recurringReservation = RecurringReservation()

        recurringReservation.booking?.bookingCode?.asString()
    }

    class RecurringReservation(val booking: Booking? = null)
}